### PR TITLE
Fix loading of charmap from cache file

### PR DESCRIPTION
**Pull request by robertknight**
Loading the cached charmap from disk always failed because the
`GzipFile` was passed to `json.loads`, which expects a string. Use
`json.load` instead, which accepts a file.

This fixes an issue where the first test in a test run that used the
charmap would be unexpectedly slow.

For context, see https://github.com/HypothesisWorks/hypothesis/issues/2108#issuecomment-548579704.

### CODE REVIEW

**Comment by Zac-HD** [comment]
Wow, that's a pretty serious bug!  Thanks heaps for spotting it, and the PR!  (with tests :heart_eyes:)

You'll need to add a release file (with thank-you note) and I'd encourage you to add yourself to the list of authors too.  See #2162 for a good example of what this looks like.

**Comment by robertknight** [comment]
> You'll need to add a release file (with thank-you note) and I'd encourage you to add yourself to the list of authors too. See #2162 for a good example of what this looks like.

Done. Thanks for the help :)

File: hypothesis-python/RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+RELEASE_TYPE: patch
+
+  This release fixes a bug (:issue:`2166`) where a Unicode character info
+  cache file was generated but never used on subsequent test runs, causing tests
+  to run more slowly than they should have.
+
+  Thanks to Robert Knight for this bugfix!
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst]: De-dent this text, and we'll be ready to merge!

(and released on PyPI ~30 minutes after that)

Comment (robertknight) [reply hypothesis-python/RELEASE.rst]: Done

**Comment by Zac-HD** [comment]
https://travis-ci.org/HypothesisWorks/hypothesis/jobs/605943000#L759-L765

**Comment by robertknight** [comment]
I'm looking into the test failure on Travis. I have a related question though about whether we could potentially just bundle a precompiled version of the charmap in the Hypothesis package, along with a script to regenerate it? The only time it changes AFAIK is when new Unicode versions are released.

On CI systems, each new branch (or sometimes every build) typically starts out with a fresh build directory, so the first test that triggers a call to `charmap()` will result in a call to this function. The caching therefore doesn't help many builds by default. The flaky test failure in our app's test suite that brought the issue to light was:

```
E   hypothesis.errors.FailedHealthCheck: Data generation is extremely slow: Only produced 6 valid examples in 1.35 seconds (0 invalid ones and 2 exceeded maximum size). Try decreasing size of the data you're generating (with e.g.max_size or max_leaves parameters).
E   See https://hypothesis.readthedocs.io/en/latest/healthchecks.html for more information about this. If you want to disable just this health check, add HealthCheck.too_slow to the suppress_health_check settings for this test.
```

**Comment by Zac-HD** [comment]
We *have* considered shipping this as a static data file, but the way that e.g. unicode version can vary by platform makes it fairly tricky to be sure that everything stays in sync (not that the status quo is perfect, either...).  Given that we want to address the 'long delay on first run' for other reasons too, we're likely to go with a more general solution.

**Comment by robertknight** [comment]
Hmm, the linux-py35 tests are failing. I'll look into that.

>  Given that we want to address the 'long delay on first run' for other reasons too, we're likely to go with a more general solution.

OK. I had a look at optimizing the existing cache generation and was able to find wins of ~2x by reducing the number of hash lookups: https://gist.github.com/robertknight/055d19c61c5dbe2c1169a54e1da7f238.

**Pull request merged**